crates/network/src/any/mod.rs

```diff
@@ -46,6 +59,480 @@ impl From<TxType> for AnyTxType {
     }
 }
 
+/// Memoization for deserialization of [`AnyTxEnvelope`] and [`AnyTypedTransaction`].
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+#[allow(unnameable_types)]
+pub struct DeserMemo {
+    input: OnceLock<Bytes>,
+    access_list: OnceLock<AccessList>,
+    blob_versioned_hashes: OnceLock<Vec<B256>>,
+    authorization_list: OnceLock<Vec<SignedAuthorization>>,
+}
+
+/// Transaction envelope for a catch-all network.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(untagged)]
+#[doc(alias = "AnyTransactionEnvelope")]
+pub enum AnyTxEnvelope {
+    /// An Ethereum transaction.
+    Ethereum(TxEnvelope),
+    /// A transaction with unknown type.
+    Other {
+        /// Transaction type.
+        #[serde(rename = "type")]
+        ty: AnyTxType,
+        /// Additional fields.
+        fields: std::collections::BTreeMap<String, serde_json::Value>,
+
+        /// Memoization for deserialization.
+        #[serde(skip, default)]
+        memo: DeserMemo,
+    },
+}
```

Comment:
I've considered such catch-all envelope variant but imo it has a couple issues
1. `fields` would consume all extra fields, leaving nothing for outer `other`. that way, if network includes some additional fields (like unusual block context or something else) for all transactions, users will need to look for them in _either_ inner `Other` variant or in outer `WithOtherFields` which is incosistent.
2. Such envelope would make it harder to access common transaction fields (like `to`, `data`, `chain_id`, etc).

In https://github.com/alloy-rs/alloy/pull/1169 I've added a `AnyTxEnvelope` type duplicating current `rpc-types-eth::Transaction` to solve those concerns. It does not have any catch-all fields and provides relatively easy access to all common fields without need to manually deserialize them. wdyt on doing something similar here?

Reply:
> fields would consume all extra fields, leaving nothing for outer other. that way, if network includes some additional fields (like unusual block context or something else) for all transactions, users will need to look for them in either inner Other variant or in outer WithOtherFields which is incosistent.

This is an unavoidable problem of having a type be a catch-all. It's an inherent problem for a serializable `Any______`. We can never distinguish between transaction data and non-transaction data in a flattened serialization, and yes it sucks, but absorbing metadata in the inner scope is better than absorbing tx data in the outer scope

> Such envelope would make it harder to access common transaction fields (like to, data, chain_id, etc).

see the implementation of the `Transaction` trait 🫡 

> In https://github.com/alloy-rs/alloy/pull/1169 I've added a AnyTxEnvelope type duplicating current rpc-types-eth::Transaction to solve those concerns. It does not have any catch-all fields and provides relatively easy access to all common fields without need to manually deserialize them. wdyt on doing something similar here?

I think it's less consistent than the style in this PR. It also doesn't capture all conceivable transaction types, e.g. ones with non-secp256k1 signatures. and a signature field of some non-`Signature` type will cause deser errors. Supporting networks with other and mixed signature types has always been a goal (actually can you review #1451 too 🙏 ). the tx envelope in #1169 supports Optimism as the trivial example (signatures may be `Signature` or `None`), but would not support hypothetical ed25519 transaction types (which is actually the motivating use case for this PR and for #1451. We implemented ed25519 as a tx type and want to try it out). You could also envision a network that change the semantics of `to` to allow named system contracts in addition to addresses and null. This would also be broken by the #1169 version

long paragraph short, I think it doesn't quite meet the goals of the network abstraction because it would conflict with transactions that change the type of transaction fields while keeping the name the same